mergify_engine/rules/parser.py

```diff
@@ -189,220 +235,268 @@ def _parse_time_range(toks: typing.List[pyparsing.Token]) -> typing.Any:
-            | non_quantifiable_attributes
-        )
-    )
-).setParseAction(_token_to_dict)
+        raise ConditionParsingError("Invalid attribute")
+
+    position += len(attribute)
+    parser = conditions_parser[attribute]
+    if parser == "bool":
+        if quantity:
+            raise ConditionParsingError(
+                f"`#` modifier is invalid for boolean attribute: `{attribute}`"
+            )
+        elif len(v[position:]) > 0:
+            raise ConditionParsingError(
+                f"Operators are invalid for boolean attribute: `{attribute}`"
+            )
+        return _to_dict(negate, False, attribute, "=", True)
+
+    elif parser == "schedule":
+        if quantity:
+            raise ConditionParsingError(
+                f"`#` modifier is invalid for attribute: `{attribute}`"
+            )
+        elif negate:
+            raise ConditionParsingError(
+                f"`-` modifier is invalid for attribute: `{attribute}`"
+            )
+        for op in equality_operators:
+            if v[position:].startswith(op):
+                break
+        else:
+            raise ConditionParsingError("Invalid operator")
+        position += len(op)
+        op = operators_mapping.get(op, op)
+        value = unquote(v[position:])
+        if op == "!=":
+            negate = True
+        cond: typing.Dict[str, typing.Any]
+        days, has_times, times = value.partition(" ")
+        try:
+            dow_cond = _extract_dowrange(days)
+        except ConditionParsingError:
+            if has_times:
+                raise
+            cond = _extract_timerange(days)
+        else:
+            if has_times:
+                time_cond = _extract_timerange(times)
+                cond = {"and": (dow_cond, time_cond)}
+            else:
+                cond = dow_cond
+        return _to_dict(negate, False, attribute, "@", cond)
+
+    elif parser == "time":
+        if quantity:
+            raise ConditionParsingError(
+                f"`#` modifier is invalid for attribute: `{attribute}`"
+            )
+        elif negate:
+            raise ConditionParsingError(
+                f"`-` modifier is invalid for attribute: `{attribute}`"
+            )
+
+        for op in range_operators:
+            if v[position:].startswith(op):
+                break
+        else:
+            raise ConditionParsingError("Invalid operator")
+        position += len(op)
+        op = operators_mapping.get(op, op)
+        value = unquote(v[position:])
+        t = _extract_time(value)
+        return _to_dict(False, False, attribute, op, t)
+
+    elif parser in ("timestamp", "timestamp_or_timedelta"):
+        if quantity:
+            raise ConditionParsingError(
+                f"`#` modifier is invalid for attribute: `{attribute}`"
+            )
+        elif negate:
+            raise ConditionParsingError(
+                f"`-` modifier is invalid for attribute: `{attribute}`"
+            )
```

Comment:
It'd be great to have a data structure that describes which "parser" allows which operation.
I'm afraid that we'll miss a case somewhere otherwise (or in the future if we had another prefix operator).

Reply:
Yeah, I copy the pyparsing parser to bench the difference. a lot of things can be factorized.